raiden/transfer/channel.py

```diff
@@ -1213,39 +1179,36 @@ def update_contract_balance(end_state: NettingChannelEndState, contract_balance:
         end_state.contract_balance = contract_balance
 
 
-def compute_merkletree_with(
-    merkletree: MerkleTreeState, lockhash: LockHash
-) -> Optional[MerkleTreeState]:
-    """Register the given lockhash with the existing merkle tree."""
-    # Use None to inform the caller the lockshash is already known
-    result = None
-
-    leaves = merkletree.layers[LEAVES]
-    if lockhash not in leaves:
-        leaves = list(leaves)
-        leaves.append(Keccak256(lockhash))
-        result = MerkleTreeState(compute_layers(leaves))
-
-    return result
+def compute_locks_with(
+    locks: PendingLocksState, lock: Union[HashTimeLockState, UnlockPartialProofState]
+) -> Optional[PendingLocksState]:
+    """Register the given lock with as a pending locks."""
+    lockhash = lock.lockhash
+    if lockhash not in locks.locks:
+        locks = PendingLocksState(dict(locks.locks))
+        locks.locks.update({lockhash: lock.encoded})  # pylint: disable=E1101
+        return locks
+    else:
+        return None
 
 
-def compute_merkletree_without(
-    merkletree: MerkleTreeState, lockhash: LockHash
-) -> Optional[MerkleTreeState]:
+def compute_locks_without(
+    locks: PendingLocksState, lockhash: LockHash
+) -> Optional[PendingLocksState]:
     # Use None to inform the caller the lockhash is unknown
-    result = None
+    if lockhash in locks.locks:
+        locks = PendingLocksState(dict(locks.locks))
+        del locks.locks[lockhash]
+        return locks
+    else:
+        return None
 
-    leaves = merkletree.layers[LEAVES]
-    if lockhash in leaves:
-        leaves = list(leaves)
-        leaves.remove(Keccak256(lockhash))
 
-        if leaves:
-            result = MerkleTreeState(compute_layers(leaves))
-        else:
-            result = make_empty_merkle_tree()
-
-    return result
+def compute_locksroot(locks: PendingLocksState) -> Locksroot:
+    """ Compute the hash representing all pending locks
+    The hash is submitted in TokenNetwork.settleChannel() call.
+    """
+    return Locksroot(keccak(b"".join(locks.locks.values())))
```

Comment:
What is the assumed order here? Are we relying on the fact that Python maintains the order of the dictionary items? Shouldn't the locks be ordered by the nonce?

I discussed this with Augusto and this can actually be a list instead of a dictionary:
1. The list order is maintained (same as the dict in this case, but we have to make it explicit here)
2. When the dictionary is converted to JSON to persist it in storage, the dictionary ordering is lost. So on restart the order we had in the dictionary is lost when the dictionary is recovered from JSON.
3. Converting to a list would mean fast O(1) access is lost, but we would be ok with having a function that: given a lockhash, it would go through pending locks and return the lock that has that hash.

Reply:
> Are we relying on the fact that Python maintains the order of the dictionary items?

Yes.

> When the dictionary is converted to JSON to persist it in storage, the dictionary ordering is lost.

That's bad.

> Converting to a list would mean fast O(1) access is lost, but we would be ok 

Yeah, up to 200 locks, it won't matter.

Reply:
I'll use a list, and stop using lockhash.